src/System.Diagnostics.TraceSource/src/System/Diagnostics/TraceInternal.cs

```diff
@@ -11,19 +11,44 @@ namespace System.Diagnostics
 {
     internal static class TraceInternal
     {
+        static TraceInternal()
+        {
+            // This is where we override default DebugProvider because we know
+            // for sure that we have some Listeners to write to.
+            Debug.SetProvider(new TraceProvider());
+        }
+
         private class TraceProvider : DebugProvider
         {
+            public override void OnIndentLevelChanged(int indentLevel)
+            {
+                if (Listeners != null)
+                {
+                    foreach (TraceListener listener in Listeners)
```

Comment:
Do we need to take the lock here instead of the `TraceListener.IndentLevel` ? If this gets called via Debug class, we would be running the enumeration without any locks taken.

Reply:
Since TraceInternal.critSec is internal and is also used in TraceListeners class (see https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.TraceSource/src/System/Diagnostics/TraceListeners.cs#L77) I made use of it here as lock.